accepted/2023/simplify-output-paths.md

```diff
@@ -0,0 +1,140 @@
+
+We are proposing this new output format because they represent an improvement over the existing behavior.  If they do represent an improvement, then we would like to change the default to use the new behavior, so that new projects and new users can get the improved behavior without having to know to opt in.  However, there are several obstacles to changing the default.
+
+First of all, we don't want to affect existing projects that haven't opted in to the new behavior.  To do otherwise would be a massive breaking change, as many CI setups, build scripts, custom build logic, etc. all depend on knowing where the output of a project is and would break if it changes.
+
+Often what we do in situations like this is to only make the new behavior the default for a new target framework.  For example, projects that hadn't explicitly opted in or out would only get the new behavior if they targeted .NET 8 or higher.  This would work for single-targeted projects, but if a project multi-targeted to .NET 7 and .NET 8, then it could be very confusing if the output for .NET 7 was in the traditional `bin` folder, but the output for .NET 8 was in an `.artifacts` folder that wasn't even in the project folder but was at the solution level.
+
+Ideally in that case we might want to use the new output path format by default for all target frameworks if any of them are .NET 8 or greater.  Unfortunately this is not possible, as the output paths need to be determined during MSBuild evaluation, and at that point it's not possible to query the build of another target framework to see if it targets a given version of .NET or higher.
+
+So for multi-targeted projects, we'd be left with these options:
+
+- Default to the new output paths - too breaking
+- Use different output path formats for different target frameworks - probably too confusing
+- Default to the old output paths
+
+Another issue with defaulting to the new output path format is that it can produce output paths that are longer, which would make it more likely for projects to run up against path length limitations.  .NET Maui projects in particular have hit this limitation in the past with nested Java builds in the intermediate output path, and has taken steps to shorten their paths.
+
+## Considerations
+
+### Using `.artifacts` for the folder name
+
+The default `.artifacts` folder name was chosen for a few reasons:
+
+- Folders starting with `.` are already ignored by the default item globs for .NET SDK projects.  This is important so that if a project switches back and forth between output formats, the output path (and especially the intermediate output path, where C# source code files are generated) won't be included in the globs.
+- The `.` prefix sorts the folder separately from the rest of the source folders
+
+There is also at least one drawback:
+
+- Folders starting with `.` are hidden in some contexts (MacOS finder for example).  This will make it harder to find the output or run the app.
```

Comment:
This seems like a big deal to me. Do we really want build output hidden by default on macOS and Linux?

Reply:
I'm not sure, but probably ideally we wouldn't want them hidden.  However, if we go with something non-hidden like `artifacts`, then we will have to add that to the list of folders that are ignored in the default item globs for ALL projects (so that you don't get weird intermediate file errors when you switch back and forth between output path formats).  That would introduce a higher breaking change risk (ie any project with source in an `artifacts` folder would now be broken).

Using a folder starting with `.` means that the default item globs already ignore it.

Reply:
I'm also very uneasy about the `.artifacts` change, it sounds like something that will be annoying for Linux/macOS users but not for Windows users, I don't think that is a good precedent.

Would it be hard to detect whether we have an existing `artifacts` folder and issue a warning in that case? Maybe that should be a standard warning anyway if you have globs matching `ArtifactsPath`.

Reply:
Sorry, what? dotfiles are a wildly inappropriate use case on Unix. This would be extremely bad. The whole point of dotfiles is that tooling goes out of their way to hide them and make them difficult to select, which is the exact opposite of what you want with build artifacts.

This signals to me that there's more work to do here. I think the SDK authors should use and review some popular Unix build tools like make, CMake, `go build`, `npm build`, `cargo build`, etc. We should review whether our output is in line with these tools, or describe how others do things differently (and why we don't want to go that route).

Reply:
Can we condition the dot based on the OS? E.g., `if Windows -> .artifacts, else -> artifacts`?

Reply:
> I think the SDK authors should use and review some popular Unix build tools like make, CMake, go build, npm run build, cargo build, etc. 

What do you think is valuable about these other builds that we should be considering here? Rather than tell people "go look at other builds", which by the way many of us have, please state instead what positive side of their output structure we should be considering here?

The intent of hiding is to match experiences other toolsets have. Specifically, the idea that scenarios like `dotnet run` would no longer pollute the workspace area with visible artifacts. That is the experience other toolsets offer that .NET does not and it's one that is a common source of confusion with customers. Other ideas were considered like putting it in `%TEMP%` but those felt like too big of a departure. 

The dotfile is not make or break for me. The centralized output is and is an enormous improvement from what we have today. Given that it's easy to tweak to dotfile, `%TEMP%`, etc ... Possible we even take scenarios like `dotnet run` and silently push artifacts to `%TEMP%` or dotfiles under the hood. 

> Can we condition the dot based on the OS?

Don't want to go that route cause it means build files, scripts, etc ... that look at build artifacts need to be conditioned on OS by default. Think it's better to have a consistent name across OS.

Reply:
> What do you think is valuable about these other builds that we should be considering here? Rather than tell people "go look at other builds", which by the way many of us have, please state instead what positive side of their output structure we should be considering here?

My current understanding is that most tools use `build` as the name of their artifacts directory. Do we want to adopt that convention?

> The intent of hiding is to match experiences other toolsets have.

I'm not aware of any tool on Unix that uses a dot prefix for actual artifacts. That is, the binary output, not intermediate files. If you're suggesting using a dot prefix for the `obj` directory, that's a different story.

Reply:
Go also takes an interesting approach by expecting multiple repos to be structured under a single "GOPATH." I think this is a pretty big departure from our current structure, but it's worth explicitly considering.

Reply:
> > What do you think is valuable about these other builds that we should be considering here? Rather than tell people "go look at other builds", which by the way many of us have, please state instead what positive side of their output structure we should be considering here?
> 
> My current understanding is that most tools use `build` as the name of their artifacts directory. Do we want to adopt that convention?
> 
> > The intent of hiding is to match experiences other toolsets have.
> 
> I'm not aware of any tool on Unix that uses a dot prefix for actual artifacts. That is, the binary output, not intermediate files. If you're suggesting using a dot prefix for the `obj` directory, that's a different story.

I should note, there are some notable exceptions like to this, like `cargo`, which appears to use `target`.

Make also doesn't have an explicit output name, but it's a pretty common convention to do `mkdir build; cd build; ../configure`.

Reply:
> What toolsets use that?

AFAIK, either by default or by convention, Make, cmake, node, and go. 

Also wanted to apologize if my comment came across as dismissive, I did not mean to suggest the authors weren't doing due diligence. But for a Unix native, I think artifacts dot files would be very abnormal.

Reply:
Github also hides the 'build' directory from search. Tbh I don't know if this is a point for or against since ideally you wouldn't be checking in your build output at all.

Reply:
> AFAIK, either by default or by convention, Make, cmake, node, and go.

Go does not use `build` that I'm aware of, it uses `bin`. For make and cmake seems more convention. 

Again though, which name we choose I'm personally less interested in. I'm not sure I agree that `build` is a clear standard but I'm not opposed to the name either. 

> Github also hides the 'build' directory from search. Tbh I don't know if this is a point for or against since ideally you wouldn't be checking in your build output at all.

If you do actually check in binary files to your repo then presumably you want to see them in GitHub. But yeah the 99.9% case is you're not checking them in so don't think this matters one way or the other.

Reply:
> Go does not use build that I'm aware of, it uses bin

Yup -- that's right, misremembered.

Reply:
> I think the really interesting thing is -- what if we named `obj` as `.obj` and stuck it inside `bin`? I suspect most people who have scripts today that reference `obj` aren't actually fetching stuff from there (they shouldn't be), they're probably just deleting it to clean up.

Is that functionally any different from the current state of the feature in Preview 3 though?

```console
$ ls .artifacts
bin  obj  package  publish
```

Reply:
> Putting .obj inside bin means you can just delete the bin directory to wipe all the build assets.

That is the critical part for me. A single top level dir for all our build assets. It's valuable enough that we've pushed that pattern into the entire .NET build stack. Want to give that goodness to our customers. 

I'm starting to come around to the idea of the top level dir isn't dot prefixed but we consider dot prefixing subdirectories which are _build implementation details_. Essentially places the 99% customer should not be peaking into.

Reply:
Anything we do to nest bin and/or obj underneath another directory is consuming precious characters from MAX_PATH on windows which is historically a real pain for Xamarin, and so now .NET MAUI apps.  There's been many changes in Android tasks/targets specifically for example to shorten the names of things we have less control over in the native toolchain/ecosystem part of the stack (eg: hashing file/folder names and building lookup tables in obj).

I'm more supportive of the shortest name possible for this reason alone.  I'd take the 6 character savings that `bin/.obj` affords over `.artifacts/obj`.

And while I understand that these paths are all customizable, what I'd like to avoid is needing to change our templates for `dotnet new maui` projects to ship with opinionated differences over the 'defaults' for other .NET project types just to ensure `dotnet build` still works out of the box on most developers' systems.

Reply:
If we keep the dot in the directory name, we should also set the Hidden attribute on Windows because all the arguments around dot in the name apply to the Hidden attribute.

Reply:
I don't like the . folder either. One good thing about the current situation is that `bin` is visible and very noticeable. This is a nice property when diving into it is required. Making the build output folder invisible is adding more "magic" to the stack.

And this is no good especially for beginners. Imagine having to explain things involving `ls -a` in the correct folder instead of just "go to the artifacts folder".